### PR TITLE
Use custom LLVM refcount pruning pass

### PR DESCRIPTION
**Pull request by sklam**
Base on https://github.com/numba/llvmlite/pull/634.

Old refcount pruning pass in Numba is a regex based implementation that works only for refcount pairs within the same basicblock.

The new custom LLVM refcount pruning pass aims to enhance the current implementation by:
- more efficient implementation in C++
- prune across basicblock:
   - diamond shape controlflow: where incref dominates decref and decref postdominates incref.
   - fanout:  incref dominates decrefs on multiple predecessor blocks.
   - fanout+raise: what fanout does + forgiving missing decref on block that raise an exception).

### CODE REVIEW

**Comment by sklam** [comment]
Currently, `NUMBA_EXPERIMENTAL_REFPRUNE_PASS` controls whether to use the LLVM refprune pass and it is default to `1` (enabled).

**Comment by sklam** [comment]
Currently failing tests:
```
numba.tests.test_array_iterators.TestArrayIterators.test_array_flat_premature_free
numba.tests.test_array_iterators.TestArrayIterators.test_array_ndenumerate_premature_free
numba.tests.test_flow_control.TestFlowControl.test_for_loop3_npm
numba.tests.test_linalg.TestBasics.test_outer
numba.tests.test_lists.TestLists.test_iteration
numba.tests.test_looplifting.TestLoopLiftingInAction.test_lift_objectmode_issue_4223
numba.tests.test_np_functions.TestNPFunctions.test_corrcoef_basic
numba.tests.test_np_functions.TestNPFunctions.test_corrcoef_edge_cases
numba.tests.test_np_functions.TestNPFunctions.test_cov_basic
numba.tests.test_np_functions.TestNPFunctions.test_interp_basic
numba.tests.test_np_functions.TestNPFunctions.test_interp_supplemental_complex_tests
numba.tests.test_np_functions.TestNPFunctions.test_interp_supplemental_tests
numba.tests.test_np_functions.TestNPFunctions.test_partition_basic
numba.tests.test_np_functions.TestNPFunctions.test_partition_boolean_inputs
numba.tests.test_np_functions.TestNPFunctions.test_partition_empty_array
numba.tests.test_np_functions.TestNPFunctions.test_partition_exception_kth_multi_dimensional
numba.tests.test_np_functions.TestNPFunctions.test_partition_fuzz
numba.tests.test_np_functions.TestNPFunctions.test_partition_iterative
numba.tests.test_np_functions.TestNPFunctions.test_partition_multi_dim
numba.tests.test_parfors.TestParfors.test_no_hoisting_with_member_function_call
numba.tests.test_parfors.TestParfors.test_two_d_array_reduction
numba.tests.test_parfors.TestParfors.test_two_d_array_reduction_reuse
numba.tests.test_parfors.TestParfors.test_two_d_array_reduction_with_float_sizes
numba.tests.test_pycc.TestCC.test_c_extension_usecase
numba.tests.test_typedlist.TestListRefctTypes.test_dict_as_item_in_list
numba.tests.test_typedlist.TestListRefctTypes.test_dict_as_item_in_list_multi_refcount
numba.tests.test_typedlist.TestListRefctTypes.test_list_as_value_in_dict
numba.tests.test_unicode_array.TestUnicodeArray.test_getitem2
```

**Comment by stuartarchibald** [comment]
I've been messing about with this: https://github.com/stuartarchibald/numba/commit/e8b43ccfd50ac5dcdff729854dbee8e42f5ce3e3 to ensure that the optimisation is stable and optimal. Without this https://github.com/numba/numba/issues/5033 still refuses to vectorize but I think this use case is potentially not, or only weakly, related.

**Comment by sklam** [comment]
All tests failing in https://github.com/numba/numba/pull/6266#issuecomment-696188669 are now passing again. Use https://github.com/numba/llvmlite/pull/634/commits/d8605c666f19b9c7f727e43ae15f7a90a651e65f.

**Comment by stuartarchibald** [comment]
xref https://github.com/numba/numba/issues/4522

**Comment by sklam** [comment]
CI is now using the `_refprune` build of llvmlite. There's a segfault for `numba.tests.test_np_functions.TestNPFunctions.test_corrcoef_edge_cases` on linux ([log](https://dev.azure.com/numba/numba/_build/results?buildId=6714&view=logs&j=f8f777fc-06bb-55d6-ff7c-3407934983ee&t=7373a89e-7181-5445-2a09-4f8529ea6efc&l=9562)).

**Comment by stuartarchibald** [comment]
This https://github.com/stuartarchibald/numba/commit/a95327778e110a332484be2b60e5997e57c329e8 adds the wiring for https://github.com/numba/llvmlite/pull/634#issuecomment-702747753

**Comment by stuartarchibald** [comment]
This adds some tests on the Numba side which assert refops are pruned as expected and also that the wiring noted above works: https://github.com/stuartarchibald/numba/commit/a1d0ce9c5d75793bc8e192bac7f63effed392a45

**Comment by sklam** [comment]
This manual CI run based on https://github.com/numba/llvmlite/pull/634/commits/dfcb6bb9f225eb9b12b4da751f058f1b08d42506 is passing: https://dev.azure.com/numba/numba/_build/results?buildId=6758&view=results

**Comment by sklam** [comment]
The latest commits added `NUMBA_EXPERIMENTAL_REFPRUNE_FLAGS` to enable specific refprune subpasses.

**Comment by sklam** [comment]
The refprune tests are too sensitive. Looks like they are dependent on platform.

File: docs/source/reference/envvars.rst [hunk 1]
```diff
@@ -311,6 +311,38 @@ Compilation options
...
     *Default value:* 128
 
+.. envvar:: NUMBA_EXPERIMENTAL_REFPRUNE_PASS
+
+    Turns on the LLVM pass level reference-count pruning pass and disables the
+    regex based implementation in Numba.
+
+    *Default value:* 1 (On)
+
+.. envvar:: NUMBA_EXPERIMENTAL_REFPRUNE_FLAGS
```
Comment (stuartarchibald) [inline docs/source/reference/envvars.rst:321]: Given it's on by default if this PR is merged as is, should it be called "experimental"?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, I've given it a review, looks good, just a few minor things to look at. Thanks again.

File: docs/source/reference/envvars.rst (hunk 1 above)
Comment (sklam) [reply docs/source/reference/envvars.rst:321]: how about `NUMBA_LLVM_REFPRUNE_FLAGS`?

Comment (stuartarchibald) [reply docs/source/reference/envvars.rst:321]: > how about `NUMBA_LLVM_REFPRUNE_FLAGS`?

sounds good.

Comment (sklam) [reply docs/source/reference/envvars.rst]: renamed

File: docs/source/reference/envvars.rst [hunk 2]
```diff
@@ -311,6 +311,38 @@ Compilation options
...
+    (case-insensitive):
+
+    - ``all``: enable all subpasses.
+    - ``per_bb``: enable per-basicblock level pruning, which is same as the
+      old regex based implementation.
+    - ``diamond``: enable inter-basicblock pruning that is a diamond shape
+      pattern, i.e. a single-entry single-exit CFG subgraph that has an incref
+      in the entry and a corresponding decref in the exit.
+    - ``fanout``: enable inter-basicblock pruning that has a fanout pattern,
+      i.e. a single-entry multiple-exit CFG subgraph that the entry has an
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply docs/source/reference/envvars.rst:336]: think its just this outstanding on the review?

File: numba/core/runtime/nrtopt.py [hunk 3]
```diff
@@ -167,6 +167,8 @@ def remove_redundant_nrt_refct(ll_module):
 
     Note: non-threadsafe due to usage of global LLVMcontext
     """
+    if config.EXPERIMENTAL_REFPRUNE_PASS:
+        return ll_module
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/runtime/nrtopt.py]: done in https://github.com/numba/numba/pull/6266/commits/6eda9fc80334ca736060a4a265d8c6b05effe493

**Review by stuartarchibald — APPROVED** [review]
Many thanks for your efforts on this, great to see it wired in! Patch looks good.

**Comment by sklam** [comment]
Needs https://github.com/numba/numba/pull/6354 to get the latest versions

**Pull request merged**